src/flow.ts

```diff
@@ -991,237 +1089,39 @@ export class Flow {
       if (this.isLocalFlag(i, LocalFlags.WRAPPED) != other.isLocalFlag(i, LocalFlags.WRAPPED)) {
         this.unsetLocalFlag(i, LocalFlags.WRAPPED); // assume not wrapped
       }
-      if (this.isLocalFlag(i, LocalFlags.NONNULL) != other.isLocalFlag(i, LocalFlags.NONNULL)) {
-        this.unsetLocalFlag(i, LocalFlags.NONNULL); // assume possibly null
-      }
+    }
+    let thisNarrowedTypes = this.narrowedTypes;
+    let otherNarrowedTypes = other.narrowedTypes;
+    if (thisNarrowedTypes && otherNarrowedTypes) {
+      thisNarrowedTypes.mergeAnd(otherNarrowedTypes);
     }
   }
 
   /** Checks if an expression of the specified type is known to be non-null, even if the type might be nullable. */
   isNonnull(expr: ExpressionRef, type: Type): bool {
     if (!type.isNullableReference) return true;
+    let thisNarrowedTypes = this.narrowedTypes;
     // below, only teeLocal/getLocal are relevant because these are the only expressions that
     // depend on a dynamic nullable state (flag = LocalFlags.NONNULL), while everything else
     // has already been handled by the nullable type check above.
     switch (getExpressionId(expr)) {
       case ExpressionId.LocalSet: {
         if (!isLocalTee(expr)) break;
         let local = this.parentFunction.localsByIndex[getLocalSetIndex(expr)];
-        return !local.type.isNullableReference || this.isLocalFlag(local.index, LocalFlags.NONNULL, false);
+        let localType = thisNarrowedTypes ? thisNarrowedTypes.get(local) : null;
```

Comment:
Can the `.get` fail here? If so, it would return `undefined` in JS and abort in Wasm. Looks like this needs another check that the `local` in fact exists?